rust/agama-cli/src/logs.rs

```diff
@@ -0,0 +1,257 @@
+fn showln(show: bool, text: &str) {
+    if !show {
+        return;
+    }
+
+    println!("{}", text);
+}
+
+// A wrapper around println which shows (or not) the text depending on the boolean variable
+fn show(show: bool, text: &str) {
+    if !show {
+        return;
+    }
+
+    print!("{}", text);
+}
+
+// Struct for log represented by a file
+struct LogPath {
+    // log source
+    src_path: String,
+
+    // directory where to collect logs
+    dst_path: PathBuf,
+}
+
+impl LogPath {
+    fn new(src: &str, dst: &Path) -> Self {
+        Self {
+            src_path: src.to_string(),
+            dst_path: dst.to_owned(),
+        }
+    }
+}
+
+// Struct for log created on demmand by a command
+struct LogCmd {
+    // command which stdout / stderr is logged
+    cmd: String,
+
+    // place where to collect logs
+    dst_path: PathBuf,
+}
+
+impl LogCmd {
+    fn new(cmd: &str, dst: &Path) -> Self {
+        Self {
+            cmd: cmd.to_string(),
+            dst_path: dst.to_owned(),
+        }
+    }
+}
+
+trait LogItem {
+    // definition of log source
+    fn from(&self) -> &String;
+
+    // definition of destination as path to a file
+    fn to(&self) -> PathBuf;
+
+    // performs whatever is needed to store logs from "from" at "to" path
+    fn store(&self) -> Result<(), Error>;
+}
+
+impl LogItem for LogPath {
+    fn from(&self) -> &String {
+        &self.src_path
+    }
+
+    fn to(&self) -> PathBuf {
+        // remove leading '/' if any from the path (reason see later)
+        let r_path = Path::new(self.src_path.as_str()).strip_prefix("/").unwrap();
+
+        // here is the reason, join overwrites the content if the joined path is absolute
+        self.dst_path.join(r_path)
+    }
+
+    fn store(&self) -> Result<(), Error> {
+        // for now keep directory structure close to the original
+        // e.g. what was in /etc will be in /<tmp dir>/etc/
+        fs::create_dir_all(self.to().parent().unwrap())?;
+
+        let options = CopyOptions::new();
+        // fs_extra's own Error doesn't implement From trait so ? operator is unusable
+        match copy_items(
+            &[self.src_path.as_str()],
+            self.to().parent().unwrap(),
+            &options,
+        ) {
+            Ok(_p) => Ok(()),
+            Err(_e) => Err(io::Error::new(
+                io::ErrorKind::Other,
+                "Copying of a file failed",
+            )),
+        }
+    }
+}
+
+impl LogItem for LogCmd {
+    fn from(&self) -> &String {
+        &self.cmd
+    }
+
+    fn to(&self) -> PathBuf {
+        self.dst_path.as_path().join(format!("{}", self.cmd))
+    }
+
+    fn store(&self) -> Result<(), Error> {
+        let cmd_parts = self.cmd.split_whitespace().collect::<Vec<&str>>();
+        let file_path = self.to();
+        let output = Command::new(cmd_parts[0])
+            .args(cmd_parts[1..].iter())
+            .output()
+            .expect("Failed to run the command");
```

Comment:
Which command? The user would get a panic but not a proper error. You might want to:

1. Return a proper error (instead of panicking). You coud use your own error type encoding more useful information
2. Or take advantage of `anyhow`'s `context` mechanism. We have other examples in the code.

Reply:
modified in a 1 way (as 2 is unknown to me yet ;-)